javascript/ql/src/semmle/javascript/security/UselessUseOfCat.qll

```diff
@@ -0,0 +1,330 @@
+      )
+  }
+
+  /**
+   * Holds if this command execution invokes the executeable `name`.
+   */
+  bindingset[name]
+  predicate isACallTo(string name) {
+    if this.hasArgumentList()
+    then getArgument(0).mayHaveStringValue(name)
+    else
+      exists(string arg | arg = getConstantStringParts(getArgument(0)) |
+        arg.prefix(name.length()) = name
+      )
+  }
+}
+
+/**
+ * Holds if the input `str` contains some character that might be interpreted in a non-trivial way by a shell. 
+ */ 
+bindingset[str]
+predicate containsNonTrivialBashChar(string str) {
+  exists(str.regexpFind("\\*|\\||>|<| |\\$|&|,|\\`| ", _, _))
+}
+
+/**
+ * Gets the constant string parts from a data-flow node.
+ * Either the result is a constant string value that the node can hold, or the node is a string-concatenation and the result is the string parts from the concatenation.
+ */
+private string getConstantStringParts(DataFlow::Node node) {
+  node.mayHaveStringValue(result)
+  or
+  result = node.(StringOps::ConcatenationRoot).getConstantStringParts()
+}
+
+/**
+ * A call to a useless use of `cat`.
+ */
+class UselessCat extends CommandCall {
+  UselessCat() {
+    this = command and
+    isACallTo(getACatExecuteable()) and
+    // There is a file to read, it's not just spawning `cat`.
+    not (
+      not exists(getArgumentList()) and
+      getArgument(0).mayHaveStringValue(getACatExecuteable())
+    ) and
+    // wildcards, pipes, redirections, other bash features, and multiple files (spaces) are OK.
+    not containsNonTrivialBashChar(getNonCommandConstantString()) and
+    // Only acceptable option is "encoding", everything else is non-trivial to emulate with fs.readFile.
+    (
+      not exists(getOptionsArg())
+      or
+      forex(string prop | exists(getOptionsArg().getALocalSource().getAPropertyWrite(prop)) |
+        prop = "encoding"
+      )
+    ) and
+    // If there is a callback, then it must either have one or two parameters, or if there is a third parameter it must be unused.
+    (
+      not exists(getCallback())
+      or
+      exists(DataFlow::FunctionNode func | func = getCallback() |
+        func.getNumParameter() = 1
+        or
+        func.getNumParameter() = 2
+        or
+        // `exec` can use 3 parameters, `readFile` can only use two, so it is OK to have a third parameter if it is unused,
+        func.getNumParameter() = 3 and
+        not exists(SSA::definition(func.getParameter(2).getParameter()))
+      )
+    ) and
+    // The process returned by an async call is unused.
+    (
+      isSync()
+      or
+      inVoidContext(this.getEnclosingExpr())
+      or
+      this.getEnclosingExpr() = any(UnusedLocal v).getAnAssignedExpr()
+    )
+  }
+}
+
+/**
+ * Gets a string used to call `cat`.
+ */
+string getACatExecuteable() {
+  result = "cat" or result = "/bin/cat"
+}
+
+/**
+ * Predicates for creating an equivalent call to `fs.readFile` from a command execution of `cat`.
+ */
+module PrettyPrintCatCall {
```

Comment:
```suggestion
module PrettyPrintCatCall {
```

Reply:
There is no change in you suggestion.

Reply:
hah. I initially added `private`, but then I realised that wouldn't work. I should have deleted the comment instead.